lib/clib/CCorresLemmas.thy

```diff
@@ -1040,76 +1040,211 @@ lemma ccorres_Guard_True_Seq:
+     apply clarsimp
+     apply (rename_tac inter_t)
+     apply (frule hoarep_exec[OF _ _ body_inv, rotated])
+      apply fastforce
+     apply (frule_tac s=inter_t in hoarep_exec[rotated 2])
+       apply fastforce
+      apply fastforce
+     apply fastforce
+    apply (metis (mono_tags, lifting) HoarePartial.SeqSwap empty_iff exec_abrupt mem_Collect_eq)
+   apply (metis (mono_tags, lifting) HoarePartial.SeqSwap exec_stuck mem_Collect_eq)
+  apply (metis (mono_tags, lifting) HoarePartial.SeqSwap empty_iff exec_fault mem_Collect_eq)
+  done
+
 lemma ccorres_While:
   assumes body_ccorres:
-    "\<And>r. ccorresG srel \<Gamma> (=) xf (G and (\<lambda>s. the (C r s))) (G' \<inter> C') hs (B r) B'"
-  and cond_ccorres:
-    "\<And>r. ccorresG srel \<Gamma> (\<lambda>rv rv'. rv = to_bool rv') cond_xf G G' hs (gets_the (C r)) setter"
-  and nf: "\<And>r. no_fail (G and (\<lambda>s. the (C r s))) (B r)"
-  and no_ofail: "\<And>r. no_ofail G (C r)"
-  and body_inv: "\<And>r. \<lbrace>G and (\<lambda>s. the (C r s))\<rbrace> B r \<lbrace>\<lambda>_. G\<rbrace>"
-                "\<Gamma> \<turnstile> (G' \<inter> C') B' G'"
-  and setter_inv_cond: "\<Gamma> \<turnstile> G' setter (G' \<inter> {s'. cond_xf s' \<noteq> 0 \<longrightarrow> s' \<in> C'})"
-  and setter_xf_inv: "\<And>val. \<Gamma> \<turnstile> {s'. xf s' = val} setter {s'. xf s' = val}"
-  shows "ccorresG srel \<Gamma> (=) xf G (G' \<inter> {s'. xf s' = r}) hs
-           (whileLoop (\<lambda>r s. the (C r s)) B r)
-           (Seq setter (While {s'. cond_xf s' \<noteq> 0} (Seq B' setter)))"
+    "\<And>r. ccorresG srel \<Gamma> rrel xf (\<lambda>s. G r s \<and> C r s = Some True) (G' \<inter> C') [] (B r) B'"
+  assumes setter_ccorres:
+    "\<And>r. ccorresG srel \<Gamma> (\<lambda>rv rv'. rv = to_bool rv') cond_xf (G r) G' [] (gets_the (C r)) setter"
+  assumes nf: "\<And>r. no_fail (\<lambda>s. G r s \<and> C r s = Some True) (B r)"
+  assumes no_ofail: "\<And>r. no_ofail (G r) (C r)"
+  assumes body_inv:
+    "\<And>r. \<lbrace>\<lambda>s. G r s \<and> C r s = Some True\<rbrace> B r \<lbrace>G\<rbrace>"
+    "\<And>r s. \<Gamma> \<turnstile> {s'. s' \<in> G' \<and> (s, s') \<in> srel \<and> G r s \<and> rrel r (xf s')
+                    \<and> s' \<in> C' \<and> C r s = Some True}
+                B' G'"
+  assumes setter_inv_cond:
+    "\<And>r s. \<Gamma> \<turnstile> {s'. s' \<in> G' \<and> (s, s') \<in> srel \<and> G r s \<and> rrel r (xf s')}
+                setter
+                {s'. s' \<in> G' \<and> (cond_xf s' \<noteq> 0 \<longrightarrow> s' \<in> C')}"
+  assumes setter_rrel:
+    "\<And>r s. \<Gamma> \<turnstile> {s'. s' \<in> G' \<and> (s, s') \<in> srel \<and> G r s \<and> rrel r (xf s')}
+                setter
+                {s'. rrel r (xf s')}"
+  shows
+    "ccorresG srel \<Gamma> rrel xf (G r) (G' \<inter> {s'. rrel r (xf s')}) hs
+       (whileLoop (\<lambda>r s. the (C r s)) B r)
+       (setter;; (While {s'. cond_xf s' \<noteq> 0} (B';; setter)))"
```

Comment:
I'm a bit puzzled by this rule (ok, besides getting dizzy trying to follow all the `*rel`s). In a "typical" C loop, you have an initialiser, a loop body with a test at the front, and then an increment/advance at the end. This pattern is so common it got enshrined as the `for` loop. In this rule, however, you have `setter; while (cond) {B; setter;}` which would indicate a much more specific type of while loop, what we'd see with a do-while loop: `do {B; setter} while (cond)`.

Is this what's happening on the C side here where you're trying to apply this rule? Did they use a do-while there? I'm not sure what the C parser does with a do-while honestly.

If that's what we're looking at, then the name `ccorres_While` might be a slightly misleading name, and possibly the C code while loop might be a bit strange.

Reply:
The `do .. while` loop potentially does one more `B` than the `while` loop, so it's not quite the same.

If I understood correctly, the `setter` part is state change that is related to the condition, but not the body, and is usually not state change that affects the state relation. A better name would be welcome, but we haven't found one yet. Come to think of it, we did talk about needing a comment for this rule, specifically describing the role of the `setter` bit. We should really add that.

You have a point that the name `ccorres_While` is a bit too general, because it does expect a specific form of while loop on the C side.

Reply:
I haven't tried looked into exactly what the C parser is doing (I doubt I would understand it anyway) but this format seems to be produced whenever there is a separate function which is used as the loop condition in a plain `while` statement. I could check, but I think when there isn't a separate function for the loop condition, then there would be no `setter` as I have in this rule.

The reason I do want to have a separate name (possibly inlined function in the C) for the loop condition is that it makes writing a `ccorres` lemma much easier than if I were to try to paste a big blob of Simpl into a `ccorres` statement. One of the `while` loops in `refill_budget_check` does not use a separate function, and the Simpl version of the loop condition there is very involved and ugly.

Would it be reasonable to always use a C function for the loop condition?

Reply:
Actually, you're right, I suspect the C parser is always going to produce this pattern if the loop condition has a potential side effect. (But will not if it doesn't). The pattern might be harder to extract if the body is not just a function call, because it'll probably just put a semicolon at the end, which is hard to match against if there are other semicolons in the body, but it should still be there in principle.

So what `setter` really is is the potentially side-effectful expression in the loop condition position in C. So a better name would actually be `condition` after all. We could look into the C parser to change associativity of that final semicolon so that the pattern here always fits for those kinds of loops, which would make the rule applicable to all while loops that have a condition with side effect.

Reply:
I would be quite happy if we always had the loop body in the C as its own function call, but I wouldn't be surprised if it's not in keeping with the style the C has. If it weren't its own function call, then could always just use one of the rules about associativity of the Simpl in a `ccorres` rule to change things around, right? I wouldn't be averse to changing what the C parser does here, though.

But thinking more generally, are we happy to always require our C `while` loops to use a standalone function as the loop condition?

Reply:
> I would be quite happy if we always had the loop body in the C as its own function call, but I wouldn't be surprised if it's not in keeping with the style the C has. 

Yes, that would definitely not be normal, having the loop body as a single function call is definitely an exception.


> If it weren't its own function call, then could always just use one of the rules about associativity of the Simpl in a `ccorres` rule to change things around, right? 

Yes, that would be possible, but somewhat annoying to use.


> I wouldn't be averse to changing what the C parser does here, though.
 
Happy to look into that (in a separately from this PR).

> But thinking more generally, are we happy to always require our C `while` loops to use a standalone function as the loop condition?

No, definitely not. The default situation is that the loop condition does not affect the state and is an otherwise arbitrary expression. Most loops are of that form. Loop conditions that do change the state (as we have here) should be the exception. Even there I don't think we can demand that the condition should always be just one function call. For the cases we are seeing in MCS, it makes sense to have them as a single function call, but I wouldn't say that in general.